flate/huffman_code.go

```diff
@@ -315,47 +313,176 @@ func (h *huffmanEncoder) generate(freq []uint16, maxBits int32) {
 		}
 		return
 	}
-	h.lfs.sort(list)
+	sortbyFreq(list)
 
 	// Get the number of literals for each bit count
 	bitCount := h.bitCounts(list, maxBits)
 	// And do the assignment
 	h.assignEncodingAndSize(bitCount, list)
 }
 
-type byLiteral []literalNode
+var sortpool = sync.Pool{New: func() interface{} {
+	return &[]literalNode{}
+}}
 
-func (s *byLiteral) sort(a []literalNode) {
-	*s = byLiteral(a)
-	sort.Sort(s)
-}
+func sortbyLiteral(list []literalNode) {
+	max_len := len(list)
+	t := sortpool.Get().(*[]literalNode)
+	defer sortpool.Put(t)
+	if len(*t) < max_len {
+		*t = make([]literalNode, max_len)
+	}
+	tmp := *t
+	for i := 0; i < max_len-max_len&1; i += 2 {
+		if list[i+1].literal < list[i].literal {
+			list[i], list[i+1] = list[i+1], list[i]
+		}
 
-func (s byLiteral) Len() int { return len(s) }
+	}
+	for i := 0; i < max_len-max_len&3; i += 4 {
+		if list[i+2].literal < list[i].literal {
+			list[i], list[i+2] = list[i+2], list[i]
+		}
+		if list[i+3].literal < list[i+1].literal {
+			list[i+1], list[i+3] = list[i+3], list[i+1]
+		}
+		if list[i+2].literal < list[i+1].literal {
+			list[i+1], list[i+2] = list[i+2], list[i+1]
+		}
 
-func (s byLiteral) Less(i, j int) bool {
-	return s[i].literal < s[j].literal
+	}
+	if max_len&3 == 3 {
+		i := max_len - 3
+		if list[i+2].literal < list[i].literal {
+			list[i+1], list[i+2] = list[i+2], list[i+1]
+			list[i], list[i+1] = list[i+1], list[i]
+		} else if list[i+2].literal < list[i+1].literal {
+			list[i+1], list[i+2] = list[i+2], list[i+1]
+		}
+	}
+	var step, l, max, r, index, n int
+	step = 4
+	for step < max_len {
+		n++
+		step <<= 1
+		if n&1 == 1 {
```

Comment:
This is a slow branch (since it is taken 50% of the time) and it seems to not really be needed with some restructure.

It will always be true on the first iteration, false on the second, etc.

Something like this would make the branching easier to predict:
```Go
	var step, l, max, r, index, n int
	step = 4
	for step < max_len {
		n++
		step <<= 1
		// Even iterations:
		for i := 0; i < max_len; i += step {
			l, r, max = i, i+step/2, i+step
			if max > max_len {
				max = max_len
			}
			for index = i; index < max; index++ {
				if l == step/2+i || (r < max && list[r].literal < list[l].literal) {
					tmp[index] = list[r]
					r++
				} else {
					tmp[index] = list[l]
					l++
				}
			}
		}

		// Odd iterations
		if step < max_len {
			n++
			step <<= 1
			for i := 0; i < max_len; i += step {
				l, r, max = i, i+step/2, i+step
				if max > max_len {
					max = max_len
				}
				for index = i; index < max; index++ {
					if l == step/2+i || (r < max && tmp[r].literal < tmp[l].literal) {
						list[index] = tmp[r]
						r++
					} else {
						list[index] = tmp[l]
						l++
					}
				}
			}
		}
	}
```

Similar below.

Reply:
it looks like better,but not faster,it must have a if.you can change it

Reply:
Because the max_len and step is a big number,the n is small number .The big number use < or >,==,it would cost more cpu time.

Reply:
You can delete 'n',no need n++. add the copy in Odd iterations else
```go
		if step < max_len {
			step <<= 1
			for i := 0; i < max_len; i += step {
				l, r, max = i, i+step/2, i+step
				if max > max_len {
					max = max_len
				}
				for index = i; index < max; index++ {
					if l == step/2+i || (r < max && tmp[r].literal < tmp[l].literal) {
						list[index] = tmp[r]
						r++
					} else {
						list[index] = tmp[l]
						l++
					}
				}
			}
		}else{
			copy(list, tmp)
		}
```